### ISSUE TITLE
Support structs

### ISSUE THREAD

**Issue by mat-hek** [membraneframework/unifex#60]
Support for structs would make it much easier to pass multiple parameters to the native functions. My proposition of the API goes as follows:

In *.spec.exs, structs could be specified like
```elixir
struct my_struct :: %My.Struct{
  id: integer,
  data: [payload],
  name: string
}
```

what would assume the existence of `My.Struct` module with the elixir struct definition, and could generate the following C struct definition:

```C
struct MyStruct {
  int id;
  UnifexPayload** data;
  char* name;
}
```

and enable passing as an argument, returning and sending the struct
```elixir
spec my_fun(my_struct) :: my_struct
sends {:my_sturct :: label, my_struct}
```

That way we could also support maps with predefined, atom keys:

```elixir
struct my_map :: %{
  id: integer,
  data: [payload],
  name: string
}
```

that would not require Elixir struct. We can also consider support for inline structs/maps:
```elixir
spec my_fun(my_struct :: %My.Struct{id: integer, data: [payload], name: string}) :: (:ok :: label)
```

that would generate

```C
struct MyFunMyStruct {
  int id;
  UnifexPayload** data;
  char* name;
}
```

cc @mickel8 @mkaput @bblaszkow06

**Comment by mickel8** [comment]
For me it sounds very cool :+1: 

Btw. I am also wondering if we should consider support for plain maps (`%{}`) here. Maybe it can have some common concepts or can be implemented in the same PR :roll_eyes:

**Comment by mat-hek** [comment]
Good point, it was already mentioned in #32. For sure we can support the case when a map has an in-advance specified set of atom keys, that could be treated just like struct but wouldn't require module and definition on the elixir side. Support for any maps is more complicated because there's no data type to reflect them in the C std lib.

**Comment by mkaput** [comment]
The proposal sound good, minor things:

- I think we should ban dots in struct names, to avoid doing some magic in name translation
- How should allocation of lists be done when returning structs?

**Comment by mat-hek** [comment]
> I think we should ban dots in struct names, to avoid doing some magic in name translation

It would enforce creating those structs in dotless modules, which would be confusing and inconvenient. What I called 'explicit naming' in the original post seems to solve that better.

> How should allocation of lists be done when returning structs?

Not sure how would it differ from allocation of any other lists...

**Comment by mat-hek** [comment]
Updated description basing on the comments above

**Comment by mickel8** [comment]
closed in #70
